Asynchronously delete small DOM nodes

NathanSolomon733 opened the pull request:
#### ef070161a369956d0787ffa9e4ae9904e27ee3cf
<pre>
Asynchronously delete small DOM nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=282249">https://bugs.webkit.org/show_bug.cgi?id=282249</a>
<a href="https://rdar.apple.com/137026148">rdar://137026148</a>

Reviewed by Matthieu Dubet and Ryosuke Niwa.

Make removedChildren DOM nodes get deleted using the Opportunistic Task
Scheduler.
Instead of synchronously deleting removed child nodes upon each call
to removeChildren, keep a strong reference to each child that has been
removed from the DOM tree but has not been deallocated. Remove the
last reference using the Opportunistic Task Scheduler which will call
each node&apos;s destructor.
This is an optimization. We see roughly a
.30% - .55% performance improvement.

* LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit-expected.txt: Added.
* LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit.html: Added.
* LayoutTests/fast/dom/gc-dom-tree-lifetime-shadow-tree-expected.txt: Added.
* LayoutTests/fast/dom/gc-dom-tree-lifetime-shadow-tree.html: Copied from LayoutTests/fast/dom/gc-dom-tree-lifetime.html.
* LayoutTests/fast/dom/gc-dom-tree-lifetime.html:
* LayoutTests/intersection-observer/root-element-deleted.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/AsyncNodeDeletionQueue.h: Added.
(WebCore::AsyncNodeDeletionQueue::addIfSubtreeSizeIsUnderLimit):
(WebCore::AsyncNodeDeletionQueue::deleteNodesNow):
(WebCore::AsyncNodeDeletionQueue::canNodeBeDeletedAsync):
(WebCore::AsyncNodeDeletionQueue::isNodeLikelyLarge):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertionMaybeAsync):
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertion):
(WebCore::ContainerNode::removeNodeWithScriptAssertion):
(WebCore::ContainerNode::replaceAll):
(WebCore::ContainerNode::removeChildren):
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::updateCanDelayNodeDeletion):
(WebCore::notifyNodeRemovedFromDocument):
(WebCore::notifyNodeRemovedFromTree):
(WebCore::notifyChildNodeRemoved):
* Source/WebCore/dom/ContainerNodeAlgorithms.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::removedLastRef):
* Source/WebCore/dom/Document.h:
(WebCore::Document::asyncNodeDeletionQueue):
* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseNoncriticalMemory):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::performOpportunisticallyScheduledTasks):
(WebCore::Page::deleteRemovedNodes):
* Source/WebCore/page/Page.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::executeOpportunisticallyScheduledTasks const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/288944@main">https://commits.webkit.org/288944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67bbefcdf1787ff5382fd684f757abbd6b8db261

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84767 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4492 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89906 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35819 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86852 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4581 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12467 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65960 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23784 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77030 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46234 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3354 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31249 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34893 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74213 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32056 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91282 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12106 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8835 "Found 1 new test failure: http/tests/security/contentSecurityPolicy/connect-src-websocket-allowed.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74435 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12333 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73561 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18221 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17941 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16384 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/3554 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17498 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11892 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15386 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13638 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4885,6 +4885,14 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
     if (m_opportunisticTaskScheduler->hasImminentlyScheduledWork())
         options.add(JSC::VM::SchedulerOptions::HasImminentlyScheduledWork);
     commonVM().performOpportunisticallyScheduledTasks(deadline, options);
+
```
LGTM
maybe we should add a comment here?

NathanSolomon733 replied:
Agreed, or wrapping the below lines in a well named function.

Andrew-Fryer commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -828,7 +828,8 @@ void ContainerNode::removeChildren()
     removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren);
-
+    for (Ref<Node>& node : removedChildren)
+        this->document().m_childrenToBeDeleted.append(RefPtr { node.ptr() });
```
I think `appendRange` might do what you want here

NathanSolomon733 replied:
Done.

NathanSolomon733 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4885,6 +4885,14 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
     commonVM().performOpportunisticallyScheduledTasks(deadline, options);
+
+    LocalFrame* localMainFrame = dynamicDowncast<LocalFrame>(mainFrame());
+    if (!localMainFrame->document())
```
I think there needs to be a check here if localMainFrame is non-null before accessing. I will amend this.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -828,11 +829,21 @@ void ContainerNode::removeChildren()
     NodeVector removedChildren;
     removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren);
-
+    delayDeletingRemovedChildren(removedChildren);
```
This is problematic. This code would mean that we'd continue to accumulate nodes in cases like this:
```js
for (let i = 0; i < 100000; ++i)
    document.body.innerHTML = '<span>some node</span>';
```

geoffreygaren replied:
I see the logic of this optimization, but I worry that there's no practical way to make it work.

When the count of nodes is small, it's OK to delay destruction, but not profitable.

When the count of nodes is large, it's profitable to delay destruction, but not OK.

geoffreygaren commented on LayoutTests/fast/dom/gc-dom-tree-lifetime-async.html:
```diff
@@ -0,0 +1,109 @@
+var testHtml = "<div id='div0-parent'><div id='div0'><div id='div0-child'></div></div><div id='div1-parent'><div id='div1'><div id='div1-child'></div></div><div id='div2-parent'><div id='div2'><div id='div2-child'></div></div></div></div></div>";
+// DOM nodes can be destructed asynchronously during idle time. Check number of nodes destructed on idle callback.
+// Use the timeout to ensure idle time occurs on the event loop after the test logic finishes so that DOM nodes will be opportunistically destructed before doing the check.
+setTimeout(() => requestIdleCallback(() => checkNumberOfNodesAlive()), 1000);
```
You can make this test faster and more reliable if you shorted the timeout and also have the callback reschedule itself if it fails, up to a maximum number of failures.

geoffreygaren commented on LayoutTests/fast/dom/gc-dom-tree-lifetime.html:
```diff
@@ -71,6 +71,7 @@
     divY = null;
     divZ = null;
     gc();
+    window.internals.forceDeleteRemovedChildNodes();
```
Since you've fixed this test to be compatible with async deletion, we don't need to copies of the test, or the complexity of a new internals.forceDeleteRemovedChildNodes() API.

geoffreygaren commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -110,9 +110,11 @@ static RemovedSubtreeObservability notifyNodeRemovedFromDocument(ContainerNode&
     for (RefPtr currentNode = &node; currentNode; currentNode = NodeTraversal::next(*currentNode)) {
         currentNode->removedFromAncestor(Node::RemovalType { /* disconnectedFromDocument */ true, treeScopeChange == TreeScopeChange::Changed }, oldParentOfRemovedTree);
+        if (isCalledFromRemoveChildren == IsCalledFromRemoveChildren::Yes)
+            node.document().childrenToBeDeleted().approximateMemoryStored += currentNode->approximateMemoryCost();
```
You're going to need stronger medicine than this. `approximateMemoryCost` does not cover any of the biggest nodes.

> ~/OpenSource/Source/WebCore/dom/Node.h:
>   149:     virtual size_t approximateMemoryCost() const { return sizeof(*this); }
> 
> ~/OpenSource/Source/WebCore/svg/SVGGraphicsElement.h:
>    68:     size_t approximateMemoryCost() const override { return sizeof(*this); }
> 
> ~/OpenSource/Source/WebCore/svg/SVGPathElement.h:
>   103:     size_t approximateMemoryCost() const final { return m_pathSegList->approximateMemoryCost(); }
> 
> ~/OpenSource/Source/WebCore/svg/SVGPathSegList.h:
>   161:     size_t approximateMemoryCost() const
> 
> ~/OpenSource/Source/WebCore/svg/SVGPolyElement.h:
>    39:     size_t approximateMemoryCost() const override;
> 
> ~/OpenSource/Source/WebCore/svg/properties/SVGAnimatedPropertyImpl.h:
>    93:     size_t approximateMemoryCost() const
> 
> 6 matches across 6 files

mdubet replied:
Aren't other parts of the codebase which depend on the (currently incomplete) implementation of `approximateMemoryCost` (GC?) ?

We should indeed fix that, but it could certainly be an independent/followup patch.

geoffreygaren reviewed (commented):
I'm still concerned that it's a really big hill to climb to get this strategy to the quality level where it can land; and much simpler to just analyze which nodes are expensive to delete and why, and fix it.

Have we taken the step to characterize which node types we're deleting on the sub-tests that speed up with this change?

mdubet commented on LayoutTests/fast/dom/gc-dom-tree-lifetime.html:
```diff
@@ -64,23 +66,37 @@
+// DOM nodes can be destructed asynchronously during idle time. Check number of nodes destructed on idle callback.
+// Use the timeout to ensure idle time occurs on the event loop after the test logic finishes so that DOM nodes will 
+// be opportunistically destructed before doing the check.
+setTimeout(() => requestIdleCallback(() => checkNumberOfNodesAlive()), 500);
```
`window.requestIdleCallback`

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,66 @@
+
+ChildrenToAsyncDelete::ChildrenToAsyncDelete() = default;
+
+void ChildrenToAsyncDelete::add(NodeVector& children)
```
We should take the vector by rvalue ref if we are gonna move it after

`NodeVector&&`

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.h:
```diff
@@ -0,0 +1,63 @@
+
+#include "ContainerNode.h"
+#include "HTMLNames.h"
+#pragma once
```
The `pragma` should be before the includes

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.h:
```diff
@@ -0,0 +1,63 @@
+
+class ChildrenToAsyncDelete {
+public:
+    ChildrenToAsyncDelete();
```
No need to declare the default ctor?

mdubet commented on Source/WebCore/dom/ContainerNode.h:
```diff
@@ -161,10 +163,10 @@ class ContainerNode : public Node {
     enum class DeferChildrenChanged : bool { No, Yes };
     enum class DidRemoveElements : bool { No, Yes };
-    DidRemoveElements removeAllChildrenWithScriptAssertion(ChildChange::Source, NodeVector& children, DeferChildrenChanged = DeferChildrenChanged::No);
+    DidRemoveElements removeAllChildrenWithScriptAssertion(ChildChange::Source, NodeVector& children, DeferChildrenChanged = DeferChildrenChanged::No, const IsCalledFromRemoveChildren = IsCalledFromRemoveChildren::No);
```
Don't need the `const`

mdubet commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4885,6 +4885,23 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
+    RefPtr document = localMainFrame ? localMainFrame->document() : nullptr;
+    if (!document)
+        return;
+    Function<void (LocalFrame &)> deleteAllRemovedChildNodes = [] (LocalFrame& frame) {
```
`auto deleteAllRemovedChildNodes`

rniwa commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,88 @@
+static HashSet<AtomString>& allowList()
+{
+    static NeverDestroyed<HashSet<AtomString>> allowList = std::initializer_list<AtomString> {
+    HTMLNames::divTag->localName(), HTMLNames::aTag->localName(),
```
We should use a switch statement with `TagName::*` instead.

rniwa commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,88 @@
+    HTMLNames::labelTag->localName(), HTMLNames::liTag->localName(), HTMLNames::markTag->localName(),
+    HTMLNames::pTag->localName(), HTMLNames::qTag->localName(), HTMLNames::spanTag->localName(),
+    HTMLNames::buttonTag->localName(), HTMLNames::ulTag->localName(), HTMLNames::brTag->localName(),
+    HTMLNames::inputTag->localName(), HTMLNames::strongTag->localName(), HTMLNames::h1Tag->localName(),
```
I bet the most of cost comes from destroying input element's shadow tree.
Maybe we can test that hypothesis by just queuing input element and see if that results in the same perf improvement?

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2007, 2014, 2015 Apple Inc. All rights reserved.
+ * Copyright (C) 2007 David Smith (catfish.man@gmail.com)
```
2024 Apple

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,153 @@
+void ChildrenToAsyncDelete::add(NodeVector&& children)
+{
+    if (!m_mustSynchronouslyDeleteChildren)
+        m_childrenToBeDeleted.appendVector(children);
```
We need to move it to appendVector :`WTFMove(children)`

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,153 @@
+namespace WebCore {
+
+bool isNodeLikelySmall(Node& node)
+{
```
Maybe we should `ASSERT(node.isElement())`  ?

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,153 @@
+
+namespace WebCore {
+
+bool isNodeLikelySmall(Node& node)
```
`const Node&`

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.h:
```diff
@@ -0,0 +1,52 @@
+class ChildrenToAsyncDelete {
+public:
+    Vector<Ref<Node>> m_childrenToBeDeleted;
+    void canNodeBeAsyncDeleted(Node&);
```
`const Node&`

mdubet commented on Source/WebCore/dom/ChildrenToAsyncDelete.h:
```diff
@@ -0,0 +1,52 @@
+    bool m_mustSynchronouslyDeleteChildren = false;
+};
+
+bool isNodeLikelySmall(Node&);
```
`const Node&`

mdubet commented on Source/WebCore/dom/ContainerNode.h:
```diff
@@ -161,10 +163,10 @@ class ContainerNode : public Node {
     bool removeNodeWithScriptAssertion(Node&, ChildChange::Source);
     ExceptionOr<void> removeSelfOrChildNodesForInsertion(Node&, NodeVector&);
-
+    void delayDeletingRemovedChildren(NodeVector& removedChildren);
```
We can move the vector `NodeVector&&`

mdubet reviewed (approved):
LGTM 👍  (wpe test failure is unrelated)

rniwa commented on Source/WebCore/dom/ChildrenToAsyncDelete.cpp:
```diff
@@ -0,0 +1,146 @@
+    ASSERT(node.isElementNode());
+    switch (downcast<Element>(node).elementName()) {
+    case NodeName::HTML_input:
+        return true;
```
You don't have to write return true for each case.
You can just aggregate cases:
```
    case NodeName::HTML_input:
    case NodeName::HTML_li:
    ...
    case NodeName::HTML_article:
        return true;
```

rniwa commented on Source/WebCore/dom/ChildrenToAsyncDelete.h:
```diff
@@ -0,0 +1,50 @@
+
+namespace WebCore {
+
+class ChildrenToAsyncDelete {
```
I don't think ChildrenToAsyncDelete is a good name.
These nodes are no longer children of any node by the time we get here.
`AsyncNodeDeletionQueue` might be a better name.

rniwa commented on Source/WebCore/dom/ChildrenToAsyncDelete.h:
```diff
@@ -0,0 +1,50 @@
+    void canNodeBeAsyncDeleted(const Node&);
+    void add(NodeVector&& children);
+    void clear();
+private:
```
Nit: Need a blank line before `private:`

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+{
+    ASSERT(node.isElementNode());
+    switch (downcast<Element>(node).elementName()) {
+    case NodeName::HTML_input:
```
Did we do an exercise of visiting popular websites and collecting what node types are used?

mdubet replied:
Improving the list will be a long running effort over time

rniwa replied:
Sure, but we should at least one round of gathering of data. I don't think relying on Speedometer alone is a wise choice here because Speedometer content is by no means representative of real world websites in terms of which HTML elements are used.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+    case NodeName::HTML_option:
+    case NodeName::HTML_cite:
+    case NodeName::HTML_mark:
+    case NodeName::HTML_attachment:
```
Attachment isn't small, is it? It represents an attachment; i.e. a file.

rniwa commented on Source/WebCore/dom/Document.h:
```diff
@@ -510,6 +511,7 @@ class Document
     WEBCORE_EXPORT DOMImplementation& implementation();
     
     Element* documentElement() const { return m_documentElement.get(); }
+    AsyncNodeDeletionQueue& nodesToAsyncDelete() { return m_nodesToAsyncDelete; };
```
We should probably call this member function: asyncNodeDeletionQueue()
and rename the member variable to m_asyncNodeDeletionQueue as well.

rniwa commented on Source/WebCore/dom/ContainerNode.h:
```diff
@@ -161,10 +163,10 @@ class ContainerNode : public Node {
     enum class DeferChildrenChanged : bool { No, Yes };
     enum class DidRemoveElements : bool { No, Yes };
-    DidRemoveElements removeAllChildrenWithScriptAssertion(ChildChange::Source, NodeVector& children, DeferChildrenChanged = DeferChildrenChanged::No);
+    DidRemoveElements removeAllChildrenWithScriptAssertion(ChildChange::Source, NodeVector& children, DeferChildrenChanged = DeferChildrenChanged::No, CanDelayNodeDeletion = CanDelayNodeDeletion::No);
```
Why do we need to use CanDelayNodeDeletion::No? What test case / scenario requires that?

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +827,26 @@ void ContainerNode::removeChildren()
     NodeVector removedChildren;
-    removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren);
-
+#if PLATFORM(MAC)
```
Why do we need this #if? opportunisticSweepingAndGarbageCollection is false in iOS so the code below will work in iOS as well as macOS.

mdubet replied:
It's for GTK platform. We should probably use `#if PLATFORM(COCOA)` instead.

rniwa replied:
But OpportunisticSweepingAndGarbageCollectionEnabled flag exists everywhere and it's off by default.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +827,26 @@ void ContainerNode::removeChildren()
+        removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren, DeferChildrenChanged::No, CanDelayNodeDeletion::Yes);
+        delayDeletingRemovedChildren(WTFMove(removedChildren));
+    } else {
+        // do not unintentionally keep a document alive.
```
What is this comment about? removeAllChildrenWithScriptAssertion is nothing to do with Document.

rniwa commented on Source/WebCore/page/Settings.yaml:
```diff
@@ -314,6 +314,14 @@ NeedsDeferKeyDownAndKeyPressTimersUntilNextEditingCommandQuirk:
       default: false
 
+OpportunisticallyDeleteNodes:
+  comment: >-
```
What does the comment mean?

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4933,6 +4933,23 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
+    RefPtr document = localMainFrame ? localMainFrame->document() : nullptr;
+    if (!document)
+        return;
+    Function<void (LocalFrame &)> deleteAllRemovedChildNodes = [] (LocalFrame& frame) {
```
No need to specify the type for Function. Just use auto.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -110,9 +110,11 @@ static RemovedSubtreeObservability notifyNodeRemovedFromDocument(ContainerNode&
     for (RefPtr currentNode = &node; currentNode; currentNode = NodeTraversal::next(*currentNode)) {
         currentNode->removedFromAncestor(Node::RemovalType { /* disconnectedFromDocument */ true, treeScopeChange == TreeScopeChange::Changed }, oldParentOfRemovedTree);
+        if (canDelayNodeDeletion == CanDelayNodeDeletion::Yes)
+            node.document().nodesToAsyncDelete().canNodeBeAsyncDeleted(*currentNode);
```
I don't think we should update document-global state like this.
We should make this function return an enum value in addition to RemovedSubtreeObservability.
e.g. this function could return `std::pair<RemovedSubtreeObservability, RemovedSubtreeDelayedDeletionType>`
where `RemovedSubtreeDelayedDeletionType` is an enum class with enum values: `Supported` and `NotSupported`.

rniwa replied:
This incurs 3! indirect loads. We shouldn't add code like that. This code is hot.

rniwa reviewed (changes requested):
There are enough issues here that I don't think we should land this PR as is.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +827,26 @@ void ContainerNode::removeChildren()
 
+void ContainerNode::delayDeletingRemovedChildren(NodeVector&& removedChildren)
+{
+    this->document().nodesToAsyncDelete().add(WTFMove(removedChildren));
```
"this->" is unnecessary here.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +827,26 @@ void ContainerNode::removeChildren()
-    removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren);
-
+#if PLATFORM(MAC)
+    if (this != &rootNode() && this->document().settings().opportunisticallyDeleteNodes() && this->document().settings().opportunisticSweepingAndGarbageCollectionEnabled()) {
```
This will incur 4 indirect loads (tree-scope -> document -> settings -> bit flag).
We should consider cache this value in Document or TreeScope at the time of construction.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+    }
+}
+
+void AsyncNodeDeletionQueue::add(NodeVector&& children)
```
This function call will occur regardless of whether m_mustSynchronouslyDeleteChildren is true or false.
That's awfully inefficient. We should probably just inline this code in the header.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+void AsyncNodeDeletionQueue::add(NodeVector&& children)
+{
+    if (!m_mustSynchronouslyDeleteChildren)
+        m_nodesToAsyncDelete.appendVector(WTFMove(children));
```
We need to limit the growth of this vector at some capacity and clear the nodes.
Otherwise, a tight loop which creates a lot of elements will cause a pathological memory allocation:
```js
for (let i = 0; i < someLargeN; ++i)
    container.innerHTML = '<div><'div>';
```

rniwa replied:
I think in order to limit the size of the nodes being kept alive, w want to count the number of nodes per each "child" being added to this `Vector`. We can gather this information in `notifyNodeRemovedFromDocument` along with the information about whether this subtree supports delayed deletion or not.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +827,26 @@ void ContainerNode::removeChildren()
-
+#if PLATFORM(MAC)
+    if (this != &rootNode() && this->document().settings().opportunisticallyDeleteNodes() && this->document().settings().opportunisticSweepingAndGarbageCollectionEnabled()) {
+        removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren, DeferChildrenChanged::No, CanDelayNodeDeletion::Yes);
```
removeAllChildrenWithScriptAssertion is also called in a bunch of other functions in this file.
Why do we not want the same optimization there?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+    if (isNodeLikelySmall(node))
+        return;
+    // a descendant in the child's tree is likely too large to delay deleting until the next idle time
+    m_mustSynchronouslyDeleteChildren = true;
```
I don't think we should have a document-global state like this.
We should have an enum state which gets updated through `notifyNodeRemovedFromDocument` as a local variable in `ContainerNode::removeAllChildrenWithScriptAssertion`.

mdubet replied:
Yes indeed we should get rid of this global state 👍

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +827,26 @@ void ContainerNode::removeChildren()
+#if PLATFORM(MAC)
+    if (this != &rootNode() && this->document().settings().opportunisticallyDeleteNodes() && this->document().settings().opportunisticSweepingAndGarbageCollectionEnabled()) {
+        removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren, DeferChildrenChanged::No, CanDelayNodeDeletion::Yes);
+        delayDeletingRemovedChildren(WTFMove(removedChildren));
```
I think the place where we want to add this optimization is `removeDetachedChildrenInContainer`.
Instead of creating a `Vector<Ref<Node>>`, we can make each DOM node refer to the next node to delete using the next sibling pointer. That's a code we used to have prior to https://commits.webkit.org/254044@main (although it had become dead code by then).

rniwa replied:
Actually, we probably need to add code here as well.
But we should really experiment with linked-list variant.
It was more efficient than using Vector at some point in the past.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+    }
+}
+
+void AsyncNodeDeletionQueue::add(NodeVector&& children, int subTreeSize)
```
surely, the tree size can't be negative so we should use `unsigned` here.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,49 @@
+
+private:
+    Vector<Ref<Node>> m_queue;
+    int m_numNodesAlive;
```
I'd call this m_nodeCount and initialize it to 0.
Also, this should be `unsigned`.

rniwa commented on Source/WebCore/dom/ContainerNode.h:
```diff
@@ -161,10 +163,15 @@ class ContainerNode : public Node {
     enum class DidRemoveElements : bool { No, Yes };
-    DidRemoveElements removeAllChildrenWithScriptAssertion(ChildChange::Source, NodeVector& children, DeferChildrenChanged = DeferChildrenChanged::No);
+    struct RemoveAllChildrenResult {
+        int subTreeSize;
```
This should be `unsinged`.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+    m_queue.appendVector(WTFMove(children));
+}
+
+bool AsyncNodeDeletionQueue::canNodeBeAsyncDeleted(const Node& node)
```
I'd call this canNodeBeDeletedAsync. That'll read more naturally.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -101,38 +101,47 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
 
     RemovedSubtreeObservability observability = RemovedSubtreeObservability::NotObservable;
+    auto childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::Yes;
+    int subTreeSize = 0;
```
`unsigned`.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -101,38 +101,47 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
+    auto childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::Yes;
+    int subTreeSize = 0;
     for (RefPtr currentNode = &node; currentNode; currentNode = NodeTraversal::next(*currentNode)) {
+        subTreeSize++;
```
We prefer `++subTreeSize` per our coding guidelines.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -101,38 +101,47 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
         updateObservability(observability, observabilityOfRemovedNode(*currentNode));
         if (RefPtr root = currentNode->shadowRoot())
-            updateObservability(observability, notifyNodeRemovedFromTree(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+            updateObservability(observability, notifyNodeRemovedFromTree(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root).removedSubtreeObservability);
```
I don't think we should ignore the number of nodes or whether nodes can be deleted async or not in shadow trees.

rniwa reviewed (changes requested):
I don't think we should ignore results for nodes in a shadow tree

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.h:
```diff
@@ -67,7 +67,14 @@ enum class RemovedSubtreeObservability : bool {
-RemovedSubtreeObservability notifyChildNodeRemoved(ContainerNode& oldParentOfRemovedTree, Node&);
+
+struct RemovedSubtreeResult {
+    int subTreeSize;
```
`unsigned`.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,110 @@
+    switch (downcast<Element>(node).elementName()) {
+    case NodeName::HTML_input:
+    case NodeName::HTML_li:
+    case NodeName::HTML_a:
```
Please sort these lexicologically.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,49 @@
+public:
+    static bool canNodeBeAsyncDeleted(const Node&);
+    void add(NodeVector&& children, int subTreeSize);
+    void clear();
```
I'd call this `deleteNodesNow` or something since it will synchronously delete the nodes.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -86,21 +86,25 @@ unsigned ScriptDisallowedScope::s_count = 0;
         bool hadElementChild = false;
+        int treeSize = 0;
         while (RefPtr child = m_firstChild.get()) {
+            treeSize++;
```
This is clearly not right. Each removed node can have more descendants.
Traversing the node tree is expensive here so we don't wanna do that either.
We should probably return 0 or `std::nullopt` for `treeSize` to signify
to the caller that we don't know how many nodes are there.

rniwa replied:
Note that we know for sure this code path wouldn't end up resulting in node removal so it's totally okay to bail out.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -142,7 +147,10 @@ ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertion(ChildChan
 
             removeBetween(nullptr, child->protectedNextSibling().get(), *child);
-            auto subtreeObservability = notifyChildNodeRemoved(*this, *child);
+            auto [subTreeSize, subtreeObservability, subtreeChildrenDeletionCanBeDelayed  ] = notifyChildNodeRemoved(*this, *child);
```
Nit: Two spaces between subtreeChildrenDeletionCanBeDelayed and ].

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -86,21 +86,25 @@ unsigned ScriptDisallowedScope::s_count = 0;
         }
         document().incDOMTreeVersion();
-        return hadElementChild ? DidRemoveElements::Yes : DidRemoveElements::No;
+        return { treeSize, hadElementChild ? DidRemoveElements::Yes : DidRemoveElements::No, childrenDeletionCanBeDelayed };
```
We didn't check whether each node can be deleted asynchronously or not either.
Maybe we need a third state for this enum class like "Unknown".

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -791,7 +802,10 @@ void ContainerNode::replaceAll(Node* node)
+
+    auto [subTreeSize, didRemoveElements, canBeDelayed ] = removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren, DeferChildrenChanged::No);
+    if (this != &rootNode() && document().isAsyncNodeDeletionEnabled() && document().isOpportunisticTaskSchedulerEnabled() && canBeDelayed == ChildrenDeletionCanBeDelayed::Yes)
+        delayDeletingRemovedChildren(WTFMove(removedChildren), subTreeSize);
```
This code will load `document()` thrice. Since `document()` involves two indirect loads, we should avoid that.

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit.html:
```diff
@@ -0,0 +1,38 @@
+<script>
+    let rootDiv = document.createElement("div");
+    document.body.appendChild(rootDiv);
+    for (let i = 0; i < 100500; ++i) {
```
No curly braces around a single line statement.

rniwa replied:
Instead of hardcoding 100500 as a magic number, can we declare a variable with a descriptive name like this:
```
const asyncNodeDeletionLimit = 100000;
```

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit.html:
```diff
@@ -0,0 +1,38 @@
+    shouldBeTrue("window.internals.numberOfLiveNodes() < 100500");
+
+    let nodeWithManyChildren = document.createElement("div");
+    for (let i = 0; i < 100500; ++i) {
```
Ditto

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit.html:
```diff
@@ -0,0 +1,38 @@
+    }
+    rootDiv.innerHTML = "";
+    gc();
+    // async node deletion queue limit is set to 100'000
```
Use 100,000 instead?

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit.html:
```diff
@@ -0,0 +1,38 @@
+    rootDiv.innerHTML = "";
+    gc();
+    // async node deletion queue limit is set to 100'000
+    shouldBeTrue("window.internals.numberOfLiveNodes() < 100500");
```
This approach is problematic in the case preceding tests had memory leaks. To make the test result not flaky, we should store the starting node count and check that the count is less than `startingValue + 100500`.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,109 @@
+    case NodeName::HTML_span:
+    case NodeName::HTML_strong:
+    case NodeName::HTML_summary:
+    case NodeName::HTML_ul:
```
Why is table, tt, u, etc... excluded? Is that because they didn't show up on any popular sites?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,109 @@
+#include "HTMLElement.h"
+#include "NodeName.h"
+
+#define MAX_SIZE_ASYNC_NODE_DELETION_QUEUE 100'000
```
What's up with the quotation mark?
We should also use `constexpr unsigned` in new code, not a macro.

mdubet replied:
(Single quote is the C++14 digit separator, to make big numbers clearer)

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,109 @@
+    m_queue.appendVector(WTFMove(children));
+}
+
+bool AsyncNodeDeletionQueue::canNodeBeDeletedAsync(const Node& node)
```
We should most definitely inline this function.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -142,7 +160,10 @@ ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertion(ChildChan
-            auto subtreeObservability = notifyChildNodeRemoved(*this, *child);
+            auto [subTreeSize, subtreeObservability, subtreeChildrenDeletionCanBeDelayed] = notifyChildNodeRemoved(*this, *child);
+            treeSize += subTreeSize;
+            if (subtreeChildrenDeletionCanBeDelayed == ChildrenDeletionCanBeDelayed::No)
```
We should probably assert that subtreeChildrenDeletionCanBeDelayed isn't unknown here.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -101,38 +101,58 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
-            updateObservability(observability, notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+        if (RefPtr root = currentNode->shadowRoot()) {
+            auto [shadowSubTreeSize, newObservability, canBeDelayed] = notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root);
+            if (canBeDelayed == ContainerNode::ChildrenDeletionCanBeDelayed::No)
```
We should probably add a helper function like updateChildrenDeletionCanBeDelayed to share code.

rniwa replied:
By the way, ChildrenDeletionCanBeDelayed is quite wordy. How about CanDelayNodeDeletion or CanDeleteNodeAsync?

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5052,6 +5052,22 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
+    if (!document)
+        return;
+    forEachLocalFrame([] (LocalFrame& frame) {
+        Document* document = frame.document();
```
RefPtr

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5052,6 +5052,22 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
+void Page::deleteRemovedChildNodes()
+{
+    auto* localMainFrame = dynamicDowncast<LocalFrame>(mainFrame());
+    RefPtr document = localMainFrame ? localMainFrame->document() : nullptr;
```
I think it's cleaner to have an early exist for when localMainFrame was nullptr.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.cpp:
```diff
@@ -0,0 +1,109 @@
+    }
+}
+
+void AsyncNodeDeletionQueue::add(NodeVector&& children, unsigned subTreeSize)
```
Can these functions be inlined?

Also, this function should probably be named addIfSubtreeSizeIsUnderLimit or something since it doesn't delay the deletion of nodes when the subtree size is big enough.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -86,11 +86,28 @@ unsigned ScriptDisallowedScope::s_count = 0;
-ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertion(ChildChange::Source source, NodeVector& children, DeferChildrenChanged deferChildrenChanged) -> DidRemoveElements
+ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertionMaybeAsync(ChildChange::Source source, NodeVector& children, DeferChildrenChanged deferChildrenChanged) -> RemoveAllChildrenResult
+{
+#if PLATFORM(MAC)
```
Why does this code need to be Mac specific? We should at least explain why in a comment or in the commit message.

rniwa replied:
Could you add a FIXME here saying that we should have a unified code path?

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -827,12 +847,16 @@ void ContainerNode::removeChildren()
     dispatchSubtreeModifiedEvent();
 }
 
+void ContainerNode::delayDeletingRemovedChildren(NodeVector&& removedChildren, unsigned subTreeSize)
```
I don't think this extra indirection is helpful. We should just directly call into async deletion queue in this function's call site instead.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -86,11 +86,28 @@ unsigned ScriptDisallowedScope::s_count = 0;
+ALWAYS_INLINE auto ContainerNode::removeAllChildrenWithScriptAssertionMaybeAsync(ChildChange::Source source, NodeVector& children, DeferChildrenChanged deferChildrenChanged) -> RemoveAllChildrenResult
+{
+#if PLATFORM(MAC)
+    if (this != &rootNode()) {
```
Why is the root node special?

rniwa commented on Source/WebCore/page/Page.h:
```diff
@@ -1181,6 +1181,7 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
 
     void opportunisticallyRunIdleCallbacks(MonotonicTime deadline);
     void performOpportunisticallyScheduledTasks(MonotonicTime deadline);
+    void deleteRemovedChildNodes();
```
I don't think "child" adds any value here. Just call it deleteRemovedNodes or maybe executeDelayedNodeDeletionNow is more clear.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+
+        switch (downcast<Element>(node).elementName()) {
+        case NodeName::HTML_a:
+        case NodeName::HTML_abbr:
```
Why is `acronym` not included?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+        switch (downcast<Element>(node).elementName()) {
+        case NodeName::HTML_a:
+        case NodeName::HTML_abbr:
+        case NodeName::HTML_address:
```
Why is `area` not included?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+        case NodeName::HTML_a:
+        case NodeName::HTML_abbr:
+        case NodeName::HTML_address:
+        case NodeName::HTML_article:
```
`aside`?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+        case NodeName::HTML_abbr:
+        case NodeName::HTML_address:
+        case NodeName::HTML_article:
+        case NodeName::HTML_b:
```
`base`, `basefont`, `bdi`, `bdo`, `big`, `blockquote`, `body`, `br`?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+        case NodeName::HTML_article:
+        case NodeName::HTML_b:
+        case NodeName::HTML_button:
+        case NodeName::HTML_cite:
```
`caption`, `center`, `code`, `col`, `colgroup`, `data`, `datalist`, `dd`, `del`?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+        return true;
+    }
+
+    ALWAYS_INLINE static bool isNodeLikelySmall(const Node& node)
```
This list is missing so many small elements. Maybe it's better to enumerate the node types that should be excluded from this optimization?

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+private:
+    Vector<Ref<Node>> m_queue;
+    unsigned m_nodeCount = 0;
+    static constexpr unsigned m_maxSizeAsyncNodeDeletionQueue = 100000;
```
Since this is a static variable, we should use `s_` prefix, not `m_`.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,125 @@
+
+private:
+    Vector<Ref<Node>> m_queue;
+    unsigned m_nodeCount = 0;
```
Nit: Use `{ 0 }` instead?

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -86,11 +86,29 @@ unsigned ScriptDisallowedScope::s_count = 0;
     ASSERT(children.isEmpty());
     collectChildNodes(*this, children);
 
+    auto canDelayNodeDeletion = CanDelayNodeDeletion::Yes;
```
It seems that we can declare this variable much later at where treeSize is declared.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -791,9 +814,8 @@ void ContainerNode::replaceAll(Node* node)
-    auto didRemoveElements = removeAllChildrenWithScriptAssertion(ChildChange::Source::API, removedChildren, DeferChildrenChanged::Yes);
-
-    auto replacedAllChildren = is<Element>(*node) || didRemoveElements == DidRemoveElements::Yes ? ReplacedAllChildren::YesIncludingElements : ReplacedAllChildren::YesNotIncludingElements;
+    auto replacedAllChildren = is<Element>(*node) || removeAllChildrenWithScriptAssertionMaybeAsync(ChildChange::Source::API, removedChildren, DeferChildrenChanged::No).didRemoveElements == DidRemoveElements::Yes
```
It seems like we should implement the same optimization here??

rniwa replied:
Please add a FIXME here to implement the same optimization.

NathanSolomon733 replied:
Isn't the optimization already being applied here since we are calling `removeAllChildrenWithScriptAssertionMaybeAsync` and not `removeAllChildrenWithScriptAssertion`?

rniwa commented on Source/WebCore/dom/ContainerNode.h:
```diff
@@ -146,6 +146,8 @@ class ContainerNode : public Node {
     ExceptionOr<void> ensurePreInsertionValidityForPhantomDocumentFragment(NodeVector& newChildren, Node* refChild = nullptr);
     ExceptionOr<void> insertChildrenBeforeWithoutPreInsertionValidityCheck(NodeVector&&, Node* nextChild = nullptr);
 
+    enum class CanDelayNodeDeletion : uint8_t { No, Yes, Unknown };
```
It seems that we can declare this at where RemoveAllChildrenResult is declared?

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -101,38 +101,58 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
-        if (RefPtr root = currentNode->shadowRoot())
-            updateObservability(observability, notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+        if (RefPtr root = currentNode->shadowRoot()) {
+            auto [shadowSubTreeSize, newObservability, canBeDelayed] = notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root);
```
I'd call the first variable `shadowTreeSize` instead since shadow tree is a pretty well established term.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
```diff
@@ -101,38 +101,58 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
-            updateObservability(observability, notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+        if (RefPtr root = currentNode->shadowRoot()) {
+            auto [shadowSubTreeSize, newObservability, canBeDelayed] = notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root);
+            if (canBeDelayed == ContainerNode::CanDelayNodeDeletion::No)
```
We should add a new helper like `updateCanDelayNodeDeletion` much like `updateObservability`.

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-lifetime.html:
```diff
@@ -71,16 +76,34 @@
+// DOM nodes can be destructed asynchronously during idle time. Check number of nodes destructed on idle callback.
+// Use the timeout to ensure idle time occurs on the event loop after the test logic finishes so that DOM nodes will 
+// be opportunistically destructed before doing the check.
+setTimeout(() => window.requestIdleCallback(checkNumberOfNodesAlive), 500);
```
500 is an awfully long wait. Can we use a shorter timeout?
Maybe we need a way to manually trigger opportunistic tasks?
e.g. `internals.executeOpportunisticallyScheduledTasks`, which executes all opportunistic tasks.
We can fallback to `setTimeout` & `requestIdleCallback` when internals is not available.
I'm afraid the current approach will make these tests flaky.

rniwa commented on Source/WebCore/dom/AsyncNodeDeletionQueue.h:
```diff
@@ -0,0 +1,90 @@
+        ASSERT(node.isElementNode());
+
+        switch (downcast<Element>(node).elementName()) {
+        case NodeName::HTML_audio:
```
Much nicer! It's probably faster too.

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5052,6 +5052,24 @@ void Page::performOpportunisticallyScheduledTasks(MonotonicTime deadline)
+
+void Page::deleteRemovedNodes()
+{
+    auto* localMainFrame = dynamicDowncast<LocalFrame>(mainFrame());
```
RefPtr?

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-lifetime-shadow-tree.html:
```diff
@@ -0,0 +1,134 @@
+<script src="../../resources/js-test-post.js"></script>
+</body>
+</html>
+
```
Nit: These trailing newlines seem unnecessary.

rniwa commented on Source/WebCore/dom/ContainerNodeAlgorithms.h:
```diff
@@ -83,4 +90,10 @@ inline void disconnectSubframesIfNeeded(ContainerNode& root, SubframeDisconnectP
     disconnectSubframes(root, policy);
 }
 
+inline void updateCanDelayNodeDeletion(ContainerNode::CanDelayNodeDeletion& currentCanDelayDeletion, ContainerNode::CanDelayNodeDeletion newStatus)
```
This doesn't need to be in the header file, does it? We can declare & define it in the cpp file instead.

rniwa commented on LayoutTests/fast/dom/gc-dom-tree-lifetime-shadow-tree.html:
```diff
@@ -0,0 +1,109 @@
+    divY = null;
+    divZ = null;
+    gc();
+    if (window.internals) {
```
When internals is not defined, we should fallback to using requestIdleCallback.
That way, this test can be run inside a browser for verification purposes.

NathanSolomon733 replied:
Because we need internals for window.internals.numberOfLiveNodes(), we can't print if nodes are still alive or not when internals is not defined, so there is nothing to execute for that case. Idle time occurs naturally if this test runs in the browser once this code finishes running. The other printing in this test will still occur though.

rniwa replied:
Oh, that's a good point.

NathanSolomon733 committed:
Asynchronously delete small DOM nodes
https://bugs.webkit.org/show_bug.cgi?id=282249
rdar://137026148

Reviewed by Matthieu Dubet and Ryosuke Niwa.

Make removedChildren DOM nodes get deleted using the Opportunistic Task
Scheduler.
Instead of synchronously deleting removed child nodes upon each call
to removeChildren, keep a strong reference to each child that has been
removed from the DOM tree but has not been deallocated. Remove the
last reference using the Opportunistic Task Scheduler which will call
each node's destructor.
This is an optimization. We see roughly a
.30% - .55% performance improvement.

* LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit-expected.txt: Added.
* LayoutTests/fast/dom/gc-dom-tree-async-node-deletion-queue-limit.html: Added.
* LayoutTests/fast/dom/gc-dom-tree-lifetime-shadow-tree-expected.txt: Added.
* LayoutTests/fast/dom/gc-dom-tree-lifetime-shadow-tree.html: Copied from LayoutTests/fast/dom/gc-dom-tree-lifetime.html.
* LayoutTests/fast/dom/gc-dom-tree-lifetime.html:
* LayoutTests/intersection-observer/root-element-deleted.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/AsyncNodeDeletionQueue.h: Added.
(WebCore::AsyncNodeDeletionQueue::addIfSubtreeSizeIsUnderLimit):
(WebCore::AsyncNodeDeletionQueue::deleteNodesNow):
(WebCore::AsyncNodeDeletionQueue::canNodeBeDeletedAsync):
(WebCore::AsyncNodeDeletionQueue::isNodeLikelyLarge):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertionMaybeAsync):
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertion):
(WebCore::ContainerNode::removeNodeWithScriptAssertion):
(WebCore::ContainerNode::replaceAll):
(WebCore::ContainerNode::removeChildren):
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::updateCanDelayNodeDeletion):
(WebCore::notifyNodeRemovedFromDocument):
(WebCore::notifyNodeRemovedFromTree):
(WebCore::notifyChildNodeRemoved):
* Source/WebCore/dom/ContainerNodeAlgorithms.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::removedLastRef):
* Source/WebCore/dom/Document.h:
(WebCore::Document::asyncNodeDeletionQueue):
* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseNoncriticalMemory):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::performOpportunisticallyScheduledTasks):
(WebCore::Page::deleteRemovedNodes):
* Source/WebCore/page/Page.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::executeOpportunisticallyScheduledTasks const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/288944@main
